### PR TITLE
Fix: MediaPlaylist with discontinuities should not ignore maps

### PR DESCRIPTION
**Pull request by vish91**
MediaPlaylist with  discontinuties should not ignore segment maps

MediaPlaylist with default Map and discontinuities should always have segment Maps, refering to the media initialization section after it ( be it content or ads  or even if it is same as the DefaultMap )

Example: 
today we get 
```#EXTM3U
#EXT-X-VERSION:5
#EXT-X-MAP:URI="https://example.com/content-init.mp4",BYTERANGE=1024000@1048576
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-TARGETDURATION:6
#EXTINF:5.000,
test01.mp4
#EXT-X-DISCONTINUITY
#EXTINF:6.000,
test02.mp4
#EXTINF:6.000,
test03.mp4
#EXT-X-DISCONTINUITY
#EXTINF:6.000,
test04.mp4
#EXTINF:6.000,
test05.mp4
```

instead we expect: 
```#EXTM3U
#EXT-X-VERSION:5
#EXT-X-MAP:URI="https://example.com/content-init.mp4",BYTERANGE=1024000@1048576
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-TARGETDURATION:6
#EXTINF:5.000,
test01.mp4
#EXT-X-DISCONTINUITY
#EXT-X-MAP:URI="https://segmentencoded.com/ad-init.mp4",BYTERANGE=1024000@1048576
#EXTINF:6.000,
test02.mp4
#EXTINF:6.000,
test03.mp4
#EXT-X-DISCONTINUITY
#EXT-X-MAP:URI="https://example.com/content-init.mp4",BYTERANGE=1024000@1048576
#EXTINF:6.000,
test04.mp4
#EXTINF:6.000,
test05.mp4
``` 

Pulled in some other changes from `grafov:master`. If looking for above changes , it should only be small change in writer.go -> https://github.com/zencoder/m3u8/pull/1/files#diff-dc03dbc53655fac7a4c8d16b1b4144c1R627-R640

### CODE REVIEW

**Review by soldiermoth — APPROVED** [review]

File: writer_test.go [hunk 1]
```diff
@@ -365,7 +365,7 @@ func TestEncodeMediaPlaylistWithDefaultMap(t *testing.T) {
 	if e != nil {
 		t.Errorf("Set map to segment failed: %s", e)
 	}
-
+	//fmt.Println(p.Encode().String())
```
Comment (bcasinclair) [inline writer_test.go:368]: Debug info?

File: writer_test.go [hunk 2]
```diff
@@ -378,6 +378,47 @@ func TestEncodeMediaPlaylistWithDefaultMap(t *testing.T) {
...
+	e = p.SetMap("https://segmentencoded.com", 1000*1024, 1024*1024)
+	if e != nil {
+		t.Errorf("Set map to segment failed: %s", e)
+	}
+
+	if e = p.Append("test03.ts", 6.0, ""); e != nil {
+		t.Errorf("Add 3nd segment to a media playlist failed: %s", e)
+	}
+	encoded := p.String()
+	//fmt.Println(p.Encode().String())
```
Comment (bcasinclair) [inline writer_test.go:411]: More debug info

File: writer_test.go (hunk 1 above)
Comment (vish91) [reply writer_test.go:368]: yeah, saw other tests same way, so assumed that was a pattern for other people coming in to get started with tests easily...

File: writer_test.go (hunk 2 above)
Comment (vish91) [reply writer_test.go:411]: same

**Pull request merged**